ooni/agent/scheduler.py

```diff
@@ -0,0 +1,409 @@
+        remaining_measurements = []
+        measurements_by_date = sorted(list_measurements(compute_size=True),
+                                      key=lambda k: k['test_start_time'])
+        for measurement in measurements_by_date:
+            if measurement['keep'] is True:
+                kept_measurements.append(measurement)
+            elif measurement['stale'] is True:
+                stale_measurements.append(measurement)
+            else:
+                remaining_measurements.append(measurement)
+
+        # This is the order in which we should begin deleting measurements.
+        ordered_measurements = (stale_measurements +
+                                remaining_measurements +
+                                kept_measurements)
+        while amount_deleted < amount_to_delete:
+            measurement = ordered_measurements.pop(0)
+            log.warn("Deleting report {0}".format(measurement["id"]))
+            measurement_path.child(measurement['id']).remove()
+            amount_deleted += measurement['size']
+
+
+class RunDeck(ScheduledTask):
+    """
+    This will run the decks that have been configured on the system as the
+    decks to run by default.
+    """
+
+    def __init__(self, director, deck_id, schedule):
+        self.deck_id = deck_id
+        self.director = director
+        # We use as identifier also the schedule time
+        identifier = 'run-deck-' + deck_id + '-' + md5(schedule).hexdigest()
+        super(RunDeck, self).__init__(schedule, identifier)
+
+    @defer.inlineCallbacks
+    def task(self):
+        deck = deck_store.get(self.deck_id)
+        yield deck.setup()
+        yield deck.run(self.director)
+
+
+class RefreshDeckList(ScheduledTask):
+    """
+    This task is configured to refresh the list of decks that are enabled.
+    """
+    identifier = 'refresh-deck-list'
+    schedule = '@hourly'
+
+    def __init__(self, scheduler, schedule=None, identifier=None):
+        self.scheduler = scheduler
+        super(RefreshDeckList, self).__init__(schedule, identifier)
+
+    def first_run(self):
+        """
+        On first run we enable the default decks.
+        """
+        for deck_id in DEFAULT_DECKS:
+            deck_store.enable(deck_id)
+
+    def task(self):
+        self.scheduler.refresh_deck_list()
+
+
+class SendHeartBeat(ScheduledTask):
+    """
+    This task is used to send a heartbeat that the probe is still alive and
+    well.
+    """
+    identifier = 'send-heartbeat'
+    schedule = '@hourly'
+
+    def task(self):
+        # XXX implement this
+        pass
+
+# Order mattters
+SYSTEM_TASKS = [
+    UpdateInputsAndResources
+]
+
+
+@defer.inlineCallbacks
+def run_system_tasks(no_input_store=False):
+    task_classes = SYSTEM_TASKS[:]
+
+    if no_input_store:
+        log.debug("Not updating the inputs")
+        task_classes.remove(UpdateInputsAndResources)
+
+    for task_class in task_classes:
+        task = task_class()
```

Comment:
This means that we can only create tasks that do not receive explicit arguments, right?